### PR TITLE
[N2PN DOCUMENTATION] New: General routing and broadcast routing

### PR DESCRIPTION
**Pull request by Ne02ptzero**
```
 _____________________________________
/ It were not best that we should all \
| think alike; it is difference of    |
| opinion that makes horse-races.     |
|                                     |
| -- Mark Twain, "Pudd'nhead Wilson's |
\ Calendar"                           /
 -------------------------------------
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||

```

### CODE REVIEW

**Comment by pgerhardt** [comment]
Propose adding a section in the header on "1.4 Motivation" to discuss reasons for needing to develop a custom protocol for this application.

**Comment by 0x5c** [comment]
So, comments time!

**Section 4.1.1** has provisions for _multi-user_ access to the nodes, notably the use of `User ID`s. Do we really want to have multi-user per node (Added complexity for both the D2N and N2N side), 
or do we want to remove it, or leave it in the RFC _"for later use"_.
**Section 4.1.1** also describes the use of the `Destination User ID` field as a `Channel` field. This limits the channel ID (the channel's name) to two bytes, which is a bit problematic for channel naming.

In **section 4.1.3**, it should be clear that:
- A node shall never ACK a _channel chat_ message
- _Channel chat_ msgs are broadcasts by nature (should be retransmitted by everyone). _Perhaps moving the channel chat to `BROADCAST`, and create a `Category` field (either `ALERT`, `weather`, or [Channel Name])? To Be Discussed_

In **section 4.2.1**, the `Issued Timestamp` field is described as the date the `BROADCAST`'s validity period begins (with possibility of it being in the future). It should be, however, the date the `BROADCAST` *was* issued by the device (or the original _Issued:_ date in the case of _gov't/weather agency_ alert relaying).

In **section 5**, paragraphs are massive and daunting to read. They should be segmented in smaller paragraphs (just add line-breaks).

In **section 5.2.1**, the use of the _exit node_ terminology might be confusing with the concept of nodes relaying information to/from the internet or other network, commonly called _exit nodes_.


I also recommend the addition of the obligation of every node to drop non-compliant packets.

**Comment by Ne02ptzero** [comment]
> **Section 4.1.1** has provisions for multi-user access to the nodes, notably the use of User IDs. Do we really want to have multi-user per node (Added complexity for both the D2N and N2N side),
or do we want to remove it, or leave it in the RFC "for later use".

I think we should keep it, and implement it. It is a pretty simple feature, easily added from the beginning.

> **Section 4.1.1** also describes the use of the Destination User ID field as a Channel field. This limits the channel ID (the channel's name) to two bytes, which is a bit problematic for channel naming.

Channel ID can be an integer identifier, and each client can keep an internal name table. The information can be covered with a `JOIN CHANNEL` packet or `NEW CHANNEL`, something like that.

> **In section 4.1.3**, it should be clear that:

>    A node shall never ACK a channel chat message
>    Channel chat msgs are broadcasts by nature (should be retransmitted by everyone). Perhaps moving the channel chat to BROADCAST, and create a Category field (either ALERT, weather, or [Channel Name])? To Be Discussed

Why channels would be Broacasts? Channel can be between X peoples, not entire network. As for the ACKs of channel message, I think a client receiving a channel chat should ack reception to the original sender. Chat message are to important to not be ACK-ed.

> In **section 4.2.1**, the Issued Timestamp field is described as the date the BROADCAST's validity period begins (with possibility of it being in the future). It should be, however, the date the BROADCAST was issued by the device (or the original Issued: date in the case of gov't/weather agency alert relaying).

Why would we need the issued date? Do we need to relay other broadcasts than ours?

> In **section 5**, paragraphs are massive and daunting to read. They should be segmented in smaller paragraphs (just add line-breaks).

Yup, i'll fix it.

> In **section 5.2.1**, the use of the exit node terminology might be confusing with the concept of nodes relaying information to/from the internet or other network, commonly called exit nodes.

Agreed too, i'll modify it.

> I also recommend the addition of the obligation of every node to drop non-compliant packets.

It was implicit in my head, i'll write it.

**Comment by 0x5c** [comment]
So the big misunderstanding: _channels_.
Channels are like forum "topics"; they are

- Public
- Listed on a page
- Created when a first post is posted [to it]

Given these points, a given channel need to:

- Exist when at-least one post is part of it
- Have it's name defined by the first post (and not a two-byte number)
- Be relayed to the whole network to ensure availability

This means that using the `Destination User ID` field for the channel ID isn't going to work. And since they have to be relayed by the whole network, they essentially are broadcasts.

**Comment by 0x5c** [comment]
The `groff` document does not produce pages. (One long document with no headers/footers)

**Review by 0x5c — APPROVED** [review]
LGTM

File: protocols/README.md [hunk 1]
```diff
@@ -0,0 +1,7 @@
+# Documentation/Protocols
+
+## Prerequisites
+
+1. [OPTIONAL/RECOMMENDED] Use a true linux/unix install. *GNUwin32* and similar software collections may not behave correctly.
+
+2. Make sure that `groff` is correctly installed. (`apt install groff` or similar) *As per issue #11*
```
Comment (JetseVerschuren) [inline protocols/README.md:7]: I would add a bit of explanation on Ubuntu half installs of groff.

**Review by JetseVerschuren — APPROVED** [review]

File: protocols/README.md (hunk 1 above)
Comment (0x5c) [reply protocols/README.md:7]: @JetseVerschuren I'll do that after the merge (should happen today or tomorrow)

**Review by JetseVerschuren — APPROVED** [review]

**Pull request merged**